### PR TITLE
JAX docs tutorial: Autodiff 101

### PR DESCRIPTION
**Pull request by 8bitmp3**
@jakevdp 

https://github.com/google/jax/issues/18585

### CODE REVIEW

File: docs/tutorials/automatic-differentiation.md [hunk 1]
```diff
@@ -1,8 +1,210 @@
...
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+JAX has a pretty general automatic differentiation (autodiff) system. Computing gradients is a critical part of modern machine learning methods, and this tutorial will walk you through a few introductory autodiff topics.
 
-For the time being, you may find some related content in the old documentation:
-- {doc}`../jax-101/04-advanced-autodiff`
-```
\ No newline at end of file
+While understanding how automatic differentiation works "under the hood" isn't crucial for using JAX in most contexts, you are encouraged to check out this quite accessible [video](https://www.youtube.com/watch?v=wG_nF1awSSY) to get a deeper sense of what's going on.
+
+And if you are looking for more advanced material, refer to the [Advanced autodiff](https://jax.readthedocs.io/en/latest/tutorials/advanced_autodiff.html) tutorial.
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:7]: Use ```{ref}`advanced-autodiff` ``` rather than the URL.

File: docs/tutorials/automatic-differentiation.md [hunk 2]
```diff
@@ -1,8 +1,210 @@
...
+
+## Gradients
+
+In JAX, you can differentiate a function with the `jax.grad` transform:
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:11]: ``` {func}`jax.grad` ```

Same elsewhere when referring to JAX APIs or modules (use `{mod}` for modules).

File: docs/tutorials/automatic-differentiation.md [hunk 3]
```diff
@@ -1,8 +1,210 @@
...
+
+```python
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:13]: Probably use `code-cell` here so the output will be embedded. This will require adding jupytext info to the header; you can copy and paste from `quickstart` or one of the other already-completed tutorials.

Same comment for other code blocks below.

File: docs/tutorials/automatic-differentiation.md [hunk 4]
```diff
@@ -1,8 +1,210 @@
...
+def loss2(params_dict):
+    preds = predict(params_dict['W'], params_dict['b'], inputs)
+    label_probs = preds * targets + (1 - preds) * (1 - targets)
+    return -jnp.sum(jnp.log(label_probs))
+
+print(grad(loss2)({'W': W, 'b': b}))
+```
+
+You can [register your own container types](https://github.com/google/jax/issues/446#issuecomment-467105048) to work with not just `jax.grad` but other JAX transformations (`jax.jit`, `jax.vmap`, and so on).
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:163]: Link to *extending pytrees* docs rather than to the issue comment

File: docs/tutorials/automatic-differentiation.md [hunk 5]
```diff
@@ -1,8 +1,210 @@
...
+JAX provides a simple convenience function that does essentially the same thing, but checks up to any order of differentiation that you like:
+
+```python
+from jax.test_util import check_grads
+check_grads(loss, (W, b), order=2)  # check up to 2nd order derivatives
+```
+
+## Next steps
+
+The [Advanced autodiff](https://jax.readthedocs.io/en/latest/tutorials/advanced_autodiff.html) tutorial provides more advanced and detailed explanations of how the ideas covered in this document are implemented in the JAX backend. Some features, such as (defining [custom derivatives](https://jax.readthedocs.io/en/latest/Custom_derivative_rules_for_Python_code.html), depend on understanding advanced automatic differentiation, so do check out the tutorial if you are interested.
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:210]: As above, prefer `{ref}` rather than URL links.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good! Some comments below.

**Comment by jakevdp** [comment]
Please resolve the conflicts at HEAD

File: docs/tutorials/automatic-differentiation.md [hunk 6]
```diff
@@ -1,9 +1,220 @@
...
+```{code-cell}
+def loss2(params_dict):
+    preds = predict(params_dict['W'], params_dict['b'], inputs)
+    label_probs = preds * targets + (1 - preds) * (1 - targets)
+    return -jnp.sum(jnp.log(label_probs))
+
+print(grad(loss2)({'W': W, 'b': b}))
+```
+
+You can [register your own container types](https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees) to work with not just {func}`jax.grad` but other JAX transformations ({func}`jax.jit`, {func}`jax.vmap`, and so on).
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:172]: Use a `{ref}` rather than a URL here

File: docs/tutorials/automatic-differentiation.md [hunk 7]
```diff
@@ -1,9 +1,220 @@
...
 
-```{note}
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+## Next steps
 
-For the time being, you may find some related content in the old documentation:
-- {doc}`../jax-101/04-advanced-autodiff`
-```
\ No newline at end of file
+The {ref}`advanced-autodiff` tutorial provides more advanced and detailed explanations of how the ideas covered in this document are implemented in the JAX backend. Some features, such as (defining [custom derivatives](https://jax.readthedocs.io/en/latest/Custom_derivative_rules_for_Python_code.html), depend on understanding advanced automatic differentiation, so do check out the tutorial if you are interested.
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:220]: We should use a `{ref}` here as well, or maybe ``` {doc}`../Custom_derivative_rules_for_Python_code` ``` since the target is a notebook and we may not want to add markdown reference targets to it.

File: docs/tutorials/automatic-differentiation.md [hunk 8]
```diff
@@ -1,9 +1,220 @@
...
+print(grad(grad(grad(jnp.tanh)))(2.0))
+```
+
+JAX's autodiff makes it easy to compute higher-order derivatives, because the functions that compute derivatives are themselves differentiable.
+
+Thus, higher-order derivatives are as easy as stacking transformations.
+
+This can be illustrated in the single-variable case:
+
+The derivative of $f(x) = x^3 + 2x^2 - 3x + 1$ can be computed as:
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:52]: Stylistically, one sentence per paragraph seems strange here.

File: docs/tutorials/automatic-differentiation.md (hunk 7 above)
Comment (8bitmp3) [reply docs/tutorials/automatic-differentiation.md]: :+1: 

```
{ref}`defining-custom-derivative-rules`
```

which will be in

```
{ref}`advanced-autodiff`
```

**Comment by 8bitmp3** [comment]
@jakevdp PTAL thank you :+1:

File: docs/tutorials/automatic-differentiation.md [hunk 9]
```diff
@@ -1,9 +1,216 @@
...
+
+JAX has a pretty general automatic differentiation (autodiff) system. Computing gradients is a critical part of modern machine learning methods, and this tutorial will walk you through a few introductory autodiff topics.
+
+While understanding how automatic differentiation works "under the hood" isn't crucial for using JAX in most contexts, you are encouraged to check out this quite accessible [video](https://www.youtube.com/watch?v=wG_nF1awSSY) to get a deeper sense of what's going on.
+
+And if you are looking for more advanced material, refer to {ref}`advanced-autodiff`.
+
+## Gradients
+
+In JAX, you can differentiate a function with the {func}`jax.grad` transform:
```
Comment (jakevdp) [inline docs/tutorials/automatic-differentiation.md:26]: nit: use "transformation" for the noun

**Review by jakevdp — APPROVED** [review]
Looks good!
